modules/search-as-you-type/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java

```diff
@@ -0,0 +1,432 @@
+
+            super(delegate.getReuseStrategy());
+            this.delegate = delegate;
+            this.withShingles = withShingles;
+            this.shingleSize = shingleSize;
+            this.withEdgeNGrams = withEdgeNGrams;
+        }
+
+        public static SearchAsYouTypeAnalyzer withShingles(Analyzer delegate, int shingleSize) {
+            return new SearchAsYouTypeAnalyzer(delegate, true, shingleSize, false);
+        }
+
+        public static SearchAsYouTypeAnalyzer withEdgeNGrams(Analyzer delegate) {
+            return new SearchAsYouTypeAnalyzer(delegate, false, -1, true);
+        }
+
+        public static SearchAsYouTypeAnalyzer withShinglesAndEdgeNGrams(Analyzer delegate, int shingleSize) {
+            return new SearchAsYouTypeAnalyzer(delegate, true, shingleSize, true);
+        }
+
+        @Override
+        protected Analyzer getWrappedAnalyzer(String fieldName) {
+            return delegate;
+        }
+
+        @Override
+        protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComponents components) {
+            // TODO we must find a way to add the last unigram term (michael jackson -> jackson)
+            TokenStream tokenStream = components.getTokenStream();
+            if (withShingles) {
+                tokenStream = new FixedShingleFilter(tokenStream, shingleSize);
+            }
+            if (withEdgeNGrams) {
+                tokenStream = new EdgeNGramTokenFilter(tokenStream, Defaults.MIN_GRAM, Defaults.MAX_GRAM, true);
+            }
+            return new TokenStreamComponents(components.getTokenizer(), tokenStream);
+        }
+
+        public boolean isWithEdgeNGrams() {
+            return withEdgeNGrams;
+        }
+
+        public int getShingleSize() {
+            return shingleSize;
+        }
+
+        public boolean isWithShingles() {
+            return withShingles;
+        }
+    }
+
+    public static class SuggesterizedFieldType extends TermBasedFieldType {
+
+        SuggesterizedFieldType(String name) {
+            setName(name);
+            setOmitNorms(true);
+            setIndexOptions(IndexOptions.DOCS_AND_FREQS_AND_POSITIONS); // todo do we need to always set omitNorms and INdexOptions
+            setTokenized(true);
+        }
+
+        SuggesterizedFieldType(SuggesterizedFieldType reference) {
+            super(reference);
+        }
+
+        @Override
+        public SuggesterizedFieldType clone() {
+            return new SuggesterizedFieldType(this);
+        }
+
+        @Override
+        public String typeName() {
+            return SuggesterizedFieldMapper.CONTENT_TYPE;
+        }
+
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            return new TermQuery(new Term(FieldNamesFieldMapper.NAME, name()));
+        }
+    }
+
+    public static class SuggesterizedFieldMapper extends FieldMapper implements ArrayValueMapperParser { // todo better name
```

Comment:
Same here, we need something to get the informations of what is indexed internally.

Reply:
We should also have a way to get these mappers (or field types) directly from the main mapper:
````
public SuggesterizedFieldType getSubType(int shingleSize, boolean withEdgeNGram);
````
?

Reply:
I added this method - I added the analysis details to the fieldtype, I think that's enough since you can do mapper.fieldType().shingleSize()